src/lib/Normalize.hs

```diff
@@ -319,17 +319,23 @@ materializeAtoms xs = do
 
 materializeAtom :: NAtom -> SimplifyM ((NAtom, [NDecl]), SimpEnv)
 materializeAtom atom = case atom of
-  NAtomicFor b body -> do
+  NAtomicFor _ _ -> do
     scope <- asks snd
     let ty = getAtomType scope atom
-    refreshBindersRSimp [b] $ \[b'] -> do
-      ((body', decls), env) <- materializeAtom body
-      scope' <- asks snd
-      let v = rename (rawName "tab") (scope' <> snd env)
-      let env' = (mempty, v @> L ty)
-      let decl = NLet [v:>ty] (NScan b' [] [] (NAtoms [body']))
-      return ((NVar v, decls ++ [decl]), env <> env')
+    atomExpr <- atomToNExpr atom
+    let v = rename (rawName "tab") scope
+    let env' = (mempty, v @> L ty)
+    let decl = NLet [v:>ty] atomExpr
+    return ((NVar v, [decl]), env')
   _ -> return ((atom, []), mempty)
+  where
+    atomToNExpr :: NAtom -> SimplifyM NExpr
+    atomToNExpr atom = case atom of
```

Comment:
GHC warns that `atom` here shadows the binding above. You could rename to `atom'` or move this function out to the top level. It's nice to try to keep the compilation warning-free because the warnings often catch real bugs.

Reply:
Yes, always better to keep warning free. I moved it to top level.